src/MatrixFields/field_name_dict.jl

```diff
@@ -24,52 +24,51 @@ matrix can be computed by calling `one(dict)`.
 
 When broadcasting over `FieldNameDict`s, the following operations are supported:
 - Addition and subtraction
-- Multiplication, where the first argument must be a `FieldMatrix` (or
-  `FieldMatrixBroadcasted`)
-- Inversion, where the argument must be a diagonal `FieldMatrix` (or
-  `FieldMatrixBroadcasted`), i.e., one in which every entry is either a
-  `ColumnwiseBandMatrixField` of `DiagonalMatrixRow`s or a multiple of
-  `LinearAlgebra.I`
+- Multiplication, where the first argument must be a `FieldMatrix`
+- Inversion, where the argument must be a diagonal `FieldMatrix`, i.e., one in
+  which every entry is either a `ColumnwiseBandMatrixField` of
+  `DiagonalMatrixRow`s or a multiple of `LinearAlgebra.I`
 """
-struct FieldNameDict{T1, T2, K <: FieldNameSet{T1}, E <: NTuple{<:Any, T2}} <:
-       AbstractDict{T1, T2}
+struct FieldNameDict{
+    T <: Union{FieldName, FieldNamePair},
+    K <: FieldNameSet{T},
+    E,
+} <: AbstractDict{T, Any}
     keys::K
     entries::E
 
     # This needs to be an inner constructor to prevent Julia from automatically
     # generating a constructor that fails Aqua.detect_unbound_args_recursively.
-    FieldNameDict{T1, T2}(
-        keys::FieldNameSet{T1, <:NTuple{N, T1}},
-        entries::NTuple{N, T2},
-    ) where {T1, T2, N} =
-        new{T1, T2, typeof(keys), typeof(entries)}(keys, entries)
+    function FieldNameDict(keys::FieldNameSet{T}, entries) where {T}
+        length(keys) == length(entries) || error(
+            "FieldNameDict cannot have different numbers of keys and entries",
+        )
+        unrolled_foreach(entries) do entry
+            check_entry(T, entry) ||
+                error("Invalid $(FieldNameDict{T}) entry: $entry")
+        end
+        return new{T, typeof(keys), typeof(entries)}(keys, entries)
+    end
 end
-function FieldNameDict{T1, T2}(
-    key_entry_pairs::Pair{<:T1, <:T2}...;
-    name_tree = nothing,
-) where {T1, T2}
+function FieldNameDict{T}(key_entry_pairs::Pair{<:T}...) where {T}
     keys = unrolled_map(pair -> pair[1], key_entry_pairs)
     entries = unrolled_map(pair -> pair[2], key_entry_pairs)
-    return FieldNameDict{T1, T2}(FieldNameSet{T1}(keys, name_tree), entries)
+    return FieldNameDict(FieldNameSet{T}(keys), entries)
 end
-FieldNameDict{T1}(args...) where {T1} = FieldNameDict{T1, Any}(args...)
-
-const FieldVectorView = FieldNameDict{FieldName, Fields.Field}
-const FieldVectorViewBroadcasted =
-    FieldNameDict{FieldName, Union{Fields.Field, Base.AbstractBroadcasted}}
-const FieldMatrix = FieldNameDict{
-    FieldNamePair,
-    Union{UniformScaling, ColumnwiseBandMatrixField},
-}
-const FieldMatrixBroadcasted = FieldNameDict{
-    FieldNamePair,
-    Union{UniformScaling, ColumnwiseBandMatrixField, Base.AbstractBroadcasted},
-}
 
-dict_type(::FieldNameDict{T1, T2}) where {T1, T2} = FieldNameDict{T1, T2}
+const FieldVectorView = FieldNameDict{FieldName}
```

Comment:
This is a UnionAll type, which could be a performance issue if used to construct instances of `FieldVectorView` (but using it for argument types / dispatch should be fine). Is this used to construct instances? If so, could we try refactoring that here? Or should that be a separate effort?

Reply:
I got rid of all the places where `FieldVectorVeiw` and `FieldMatrix` were being used to construct instances in this PR. They are only used for dispatch now. Getting rid of the places where they were being used to construct instances ended up fixing the allocations in the prognostic EDMF unit test.